src/js/libgeo/GeoOps.js

```diff
@@ -625,6 +625,144 @@ geoOps.PointOnSegment.updatePosition = function(el) {
 };
 geoOps.PointOnSegment.stateSize = 2;
 
+geoOps._helper.PointOnArcCr = function(arc,P){
+    var A = arc.startPoint, B = arc.viaPoint, C = arc.endPoint;
+    //var cr = List.crossratio3harm(P,A,B,C,List.ii);
+    var cr = List.crossratio3harm(A,C,B,P,List.ii);
+    var m = cr.value[0];
+    var n = cr.value[1];
+    if (!CSNumber._helper.isAlmostZero(m)) {
+        n = CSNumber.div(n, m);
+        m = CSNumber.real(1);
+    } else {
+        m = CSNumber.div(m, n);
+        n = CSNumber.real(1);
+    }
+    return [m,n];
+};
+
+geoOps._helper.projectPointToArc = function(arc,P){
+    var conic = arc.matrix;
+    var cen = geoOps._helper.CenterOfConic(conic);
+
+    var diam = List.cross(cen, P);
+ //   Render2D.drawline(diam);
+//    debugger;
+
+    var isec = geoOps._helper.IntersectLC(diam, arc.matrix);
+
+
+    var crA = geoOps._helper.PointOnArcCr(arc, isec[0]);
+    var prodA = CSNumber.mult(crA[0], crA[1]);
+   // console.log("pA",niceprint(prodA));
+    if (crA[0].value.real < 0) prodA = CSNumber.neg(prodA);
+
+    var crB = geoOps._helper.PointOnArcCr(arc,isec[1]);
+    var prodB = CSNumber.mult(crB[0], crB[1]);
+//    console.log("pB",niceprint(prodB));
+    if (crB[0].value.real < 0) prodB = CSNumber.neg(prodB);
+
+    
+    var erg;
+    if(prodA.value.real > 0 && prodB.value.real > 0){
+        console.log("hit1");
+        var d1 = List.projectiveDistMinScal(P, isec[0]);
+        var d2 = List.projectiveDistMinScal(P, isec[1]);
+
+        erg = d1 < d2 ? isec[0] : isec[1];
+    }
+    else if (prodA.value.real < 0 && prodB.value.real < 0){
+        console.log("hit rand proj");
+    var d1 = List.projectiveDistMinScal(pos, arc.startPoint);
+    var d2 = List.projectiveDistMinScal(pos, arc.endPoint);
+
+    erg = d1 < d2 ? arc.startPoint : arc.endPoint;
+    }
+    else if(prodA.value.real < 0){ 
+        console.log("hit3");
+        erg = isec[1];
+    }
+    else //(prodB.value.real < 0)
+    {
+        console.log("hit4");
+        erg = isec[0];
+    }
+
+
+    return List.normalizeMax(erg);
+};
+
+
+geoOps.PointOnArc = {};
+geoOps.PointOnArc.kind = "P";
+geoOps.PointOnArc.signature = ["C"];
+geoOps.PointOnArc.isMovable = true;
+geoOps.PointOnArc.initialize = function(el) {
+    var arc = csgeo.csnames[el.args[0]];
```

Comment:
A `signatureConstraint` is a function which can perform arbitrary checks. We currently [use this for `OtherPointOnCircle`](https://github.com/CindyJS/CindyJS/blob/v0.7.3/src/js/libgeo/GeoOps.js#L577-L579) which only makes sense for circles. But we could add a check for the `isArc` flag. Come to think of it, we should add that same check to [`TransformArc`](https://github.com/CindyJS/CindyJS/blob/v0.7.3/src/js/libgeo/GeoOps.js#L2466-L2475) which has the same kind of problem.

Personally I don't like this whole “types are strings” business here, since it doesn't capture inheritance. Every segment is a line. Every circle is a conic. Every arc is a circle. At least in our world. But look at all the places where we have to either make sure to treal lines and segments alike, or check whether a given conic is a circle and recover if it is not. But I think if we are going to change this, it will be in a later commit and on a larger scale, so for now signature constraints are the way to go, I think.